aws_lambda_powertools/utilities/streaming/s3_object.py

```diff
@@ -0,0 +1,235 @@
+        gunzip: Optional[bool] = False,
+        csv: Optional[bool] = False,
+    ):
+        self.bucket = bucket
+        self.key = key
+        self.version_id = version_id
+
+        # The underlying seekable IO, where all the magic happens
+        self.raw_stream = _S3SeekableIO(
+            bucket=bucket, key=key, version_id=version_id, boto3_s3_resource=boto3_s3_resource
+        )
+
+        # Stores the list of data transformations
+        self._data_transformations: List[BaseTransform] = []
+        if gunzip:
+            self._data_transformations.append(GzipTransform())
+        if csv:
+            self._data_transformations.append(CsvTransform())
+
+        # Stores the cached transformed stream
+        self._transformed_stream: Optional[IO[bytes]] = None
+
+    @property
+    def size(self) -> int:
+        """
+        Retrieves the size of the underlying S3 object
+        """
+        return self.raw_stream.size
+
+    @property
+    def transformed_stream(self) -> IO[bytes]:
+        """
+        Returns a IO[bytes] stream with all the data transformations applied in order
+        """
+        if self._transformed_stream is None:
+            # Apply all the transformations
+            transformed_stream = self.raw_stream
+            for transformation in self._data_transformations:
+                transformed_stream = transformation.transform(transformed_stream)
+
+            self._transformed_stream = transformed_stream
+
+        return self._transformed_stream
+
+    @overload
+    def transform(
+        self, transformations: Union[BaseTransform[T], Sequence[BaseTransform[T]]], in_place: Literal[True]
+    ) -> T:
+        pass
+
+    @overload
+    def transform(
+        self, transformations: Union[BaseTransform[T], Sequence[BaseTransform[T]]], in_place: Literal[False]
+    ) -> None:
+        pass
+
+    @overload
+    def transform(self, transformations: Union[BaseTransform[T], Sequence[BaseTransform[T]]]) -> T:
+        pass
+
+    def transform(
+        self, transformations: Union[BaseTransform[T], Sequence[BaseTransform[T]]], in_place: Optional[bool] = False
+    ) -> Optional[T]:
+        """
+        Applies one or more data transformations to the stream.
+
+        Parameters
+        ----------
+        transformations: BaseTransform[T] | Sequence[BaseTransform[T]]
+            One or more transformations to apply. Transformations are applied in the same order as they are declared.
+        in_place: bool, optional
+            Transforms the stream in place, instead of returning a new stream object. Defaults to false.
+
+        Returns
+        -------
+        T[bound=IO[bytes]], optional
+            If in_place is False, returns an IO[bytes] object representing the transformed stream
+        """
+        if self.tell() != 0:
+            raise ValueError(f"Cannot add transformations to a read object. Already read {self.tell()} bytes")
```

Comment:
As we have seek, is there any action they could do if they truly want it? if so, add at the end for an additionally helpful error.

Reply:
@heitorlessa The idea here is that once you start reading from a stream, you should not change the data transformations. We want to avoid the programming error of:

```
s3.transform(GzipTransform())
s3.readline()
s3.transform(CsvTransform())
```

which doesn't much sense, and has nothing to do with the possibility of seeking on the stream.

Do you have a better way to encode this guard?